### PR TITLE
fix: use default template if timestamp is empty (undefined) in InfluxDB

### PR DESCRIPTION
**Pull request by SergeTupchiy**
Closes https://emqx.atlassian.net/browse/EMQX-8926

### CODE REVIEW

File: lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl [hunk 1]
```diff
@@ -416,14 +418,8 @@ to_config([Item0 | Rest], Acc, Precision) ->
...
-preproc_tmpl_timestamp(undefined, <<"ns">>) ->
-    erlang:system_time(nanosecond);
-preproc_tmpl_timestamp(undefined, <<"us">>) ->
-    erlang:system_time(microsecond);
-preproc_tmpl_timestamp(undefined, <<"ms">>) ->
-    erlang:system_time(millisecond);
-preproc_tmpl_timestamp(undefined, <<"s">>) ->
-    erlang:system_time(second);
+preproc_tmpl_timestamp(undefined, Precesion) ->
+    preproc_tmpl_timestamp(?DEFAULT_TIMESTAMP_TMPL, Precesion);
```
Comment (id) [inline lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl:422]: Nit
```suggestion
    preproc_tmpl_timestamp(?DEFAULT_TIMESTAMP_TMPL, Precision);
```

Comment (SergeTupchiy) [reply lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl:422]: :facepalm: Thanks a lot, fixed

File: lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl [hunk 2]
```diff
@@ -416,14 +418,8 @@ to_config([Item0 | Rest], Acc, Precision) ->
...
-preproc_tmpl_timestamp(undefined, <<"ns">>) ->
-    erlang:system_time(nanosecond);
-preproc_tmpl_timestamp(undefined, <<"us">>) ->
-    erlang:system_time(microsecond);
-preproc_tmpl_timestamp(undefined, <<"ms">>) ->
-    erlang:system_time(millisecond);
-preproc_tmpl_timestamp(undefined, <<"s">>) ->
-    erlang:system_time(second);
+preproc_tmpl_timestamp(undefined, Precision) ->
+    preproc_tmpl_timestamp(?DEFAULT_TIMESTAMP_TMPL, Precision);
```
[the remark this answers is not in the document]
Comment (zmstone) [reply lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl:422]: I'm afraid we'll have to force all inputs to be at milliseconds level,
and perform a conversion in `parse_timestamp/1` to convert to the desired precision towards influxdb.

[the remark this answers is not in the document]
Comment (zmstone) [reply lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl:422]: or maybe influxdb client does the conversion ?

[the remark this answers is not in the document]
Comment (zmstone) [reply lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl:422]: nope:
```
  encode_(Point = #{measurement := Measurement, fields := Fields}) ->
      [encode_measurement(Measurement),
        encode_tags(maps:get(tags, Point, #{})),
        " ", encode_fields(Fields),
        case maps:get(timestamp, Point, undefined) of
            undefined -> [];
            Timestamp -> [" ", encode_timestamp(Timestamp)]
        end,
        "\n"];

```
and
```
  encode_timestamp(Timestamp) when is_integer(Timestamp) ->
      erlang:integer_to_binary(Timestamp).
```

[the remark this answers is not in the document]
Comment (SergeTupchiy) [reply lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl]: I've added a fix to use ms precision if timestamp is undefined or it is ${timestamp}.

File: lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl [hunk 3]
```diff
@@ -287,14 +287,24 @@ client_config(
...
     [
         {host, str(Host)},
         {port, Port},
         {pool_size, erlang:system_info(schedulers)},
         {pool, InstId},
-        {precision, atom_to_binary(maps:get(precision, Config, ms), utf8)}
+        {precision, atom_to_binary(Precision, utf8)}
     ] ++ protocol_config(Config).
 
+precision(_Precision0, #{write_syntax := [#{timestamp := Ts} | _]}) when
```
Comment (zmstone) [inline lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl:299]: we should probably not take the first line as precision for the entire client.

**Review by savonarola — DISMISSED** [review]

**Comment by qzhuyan** [comment]
nit, Is a changelog needed here?

**Review by qzhuyan — DISMISSED** [review]

File: changes/ee/fix-10087.zh.md [hunk 4]
```diff
@@ -0,0 +1 @@
+在InfluxDB中插入数据时，如果时间戳为空（未定义），使用默认模板。在此变化之前，如果时间戳字段没有设置，InfluxDB桥不会插入新数据。
```
Comment (id) [inline changes/ee/fix-10087.zh.md:1]: @qzhuyan could you check this please?

Comment (zmstone) [inline changes/ee/fix-10087.zh.md:1]:
```suggestion
在 InfluxDB 中插入数据时，如果时间戳为空（未定义），则使用默认的占位符 `${timestamp}`。
在此修复前，如果时间戳字段没有设置，InfluxDB 桥接使用了一个错误的时间戳。
```

Comment (qzhuyan) [inline changes/ee/fix-10087.zh.md:1]:
```suggestion
在 InfluxDB 中插入数据时，如果时间戳为空（未定义），使用默认模板。在此修复之前，如果时间戳字段没有设置，InfluxDB 桥不会插入新数据。
```

File: changes/ee/fix-10087.en.md [hunk 5]
```diff
@@ -0,0 +1 @@
+Use default template if timestamp is empty (undefined) when inserting data in InfluxDB. Prior to this change InfluxDB bridge was not inserting new data if timestamp field was not set.
```
Comment (zmstone) [inline changes/ee/fix-10087.en.md:1]:
```suggestion
Use default template `${timestamp}` if the `timestamp` config is empty (undefined) when inserting data in InfluxDB. 
Prior to this change, InfluxDB bridge inserted a wrong timestamp when template is not provided.
```

File: lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl [hunk 6]
```diff
@@ -232,15 +236,14 @@ do_start_client(
     ClientConfig,
     Config = #{write_syntax := Lines}
 ) ->
+    Precision = maps:get(precision, Config, ms),
```
Comment (zmstone) [inline lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl:239]: Question:
is Precision here an atom like `ns` or binary like `<<"ns">>` ?

Comment (SergeTupchiy) [reply lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl:239]: atom

Comment (SergeTupchiy) [reply lib-ee/emqx_ee_connector/src/emqx_ee_connector_influxdb.erl:239]: It's converted to binary in `ClientConfig`, but `Config' holds atoms, I l've verified it with manual test and tracing.

**Review by zmstone — APPROVED** [review]

**Review by thalesmg — APPROVED** [review]
:heavy_check_mark: 

Not needed now, but maybe it'd be nice to have test cases where: i) we set the timestamp template as a fixed number; ii) we check that the timestamp is not fixed anymore (the `assert_persisted_data` doesn't check that I believe).

**Comment by zmstone** [comment]
> ✔️
> 
> Not needed now, but maybe it'd be nice to have test cases where: i) we set the timestamp template as a fixed number; ii) we check that the timestamp is not fixed anymore (the `assert_persisted_data` doesn't check that I believe).

i) there is a test case for it.

**Pull request merged**